### ISSUE TITLE
Allow PlayMusic() to work from Sound Library / move sound code to SDL?

### ISSUE THREAD

**Issue by droscoe** [vpinball/vpinball#759]
**Is your feature request related to a problem? Please describe.**
The behavior of PlayMusic() is ideal for adding features like an in-game jukebox.  It is mutually exclusive in that playing another song automatically ends the previous and, it has a handy callback in the table script to process the event when a song ends.  Unfortunately, it only works for external music files. 

**Describe the solution you'd like**
I am suggesting adding the ability of using PlayMusic() for songs stored in the Sound Library.  This gives more flexibility.  For example, if I want to include some default songs without having the user download and install a supplemental music file to store in VPinball/music.  Not supplying a path defaults to using the root of vpinball/music, so simply omitting the path is not an option.  It would likely require a flag indicating internal or external storage in the command

**Describe alternatives you've considered**
I've tried PlaySound, but it does not automatically stop previous sounds, nor has the option to.  It also does not have a way to determine when a song has finished playing, which is a necessary feature for a jukebox type application

**Additional context**

**Comment by droscoe** [comment]
Looking at the code, it appears that PlayMusic() uses BASS Audio Library and PlaySound() uses Windows PlaySoundA/W.  I'm guessing that this has access to the table container whereas PlayMusic() does not?  I imagine then, that this request is not possible to implement?

**Comment by toxieainc** [comment]
It would also be possible, by passing in the in-mem data instead of the file.

But lets step back a bit, cause in general there is (at least) one thing that should be done before all of that:

There is the ONLY_USE_BASS define that can be used to get rid of the old windows API based sound playing, that i started years ago.
Unfortunately its not fully compatible yet, see e.g. https://github.com/vpinball/vpinball/issues/224 and https://github.com/vpinball/vpinball/issues/572 and https://github.com/vpinball/vpinball/issues/566 .

One could even be more radical and consider a fully OSS based (and still well maintained and multi-platform!) sound lib instead of BASS (i guess SDL already has plenty of that ;)), too, while at it, and base the whole of the sound code on that.

Would be great if you could pick this up..  :)

**Comment by jsm174** [comment]
FWIW, I think https://github.com/vpinball/vpinball/issues/572 might be fixed in standalone (but not backported to master), so it would be interesting to try `ONLY_USE_BASS` again on windows. 

I mentioned this on discord, but Blood Machines uses the WMP ocx which we were able to recreate in standalone. I've added all the methods in `AudioPlayer` to accomplish it. Also, I added methods for audio streams. I needed that to get libpinmame's audio buffer playing in standalone. 

https://github.com/vpinball/vpinball/blob/cb17c1d9e6b7d2e84109221331ae0d662a3d703d/audioplayer.cpp#L198-L299

https://github.com/vpinball/vpinball/blob/cb17c1d9e6b7d2e84109221331ae0d662a3d703d/standalone/inc/wmp/WMPControls.cpp#L15-L59

**Comment by vbousquet** [comment]
If we go the SDL way, we could choose to use with https://github.com/libsdl-org/SDL_mixer which seems to offer everything we need. It would be really nice to drop the legacy windows only stuff.

**Comment by jsm174** [comment]
I would be all for that. I have to do a bunch some trickery (approved) on tvOS with BASS due to licensing issues.

**Comment by toxieainc** [comment]
> FWIW, I think #572 might be fixed in standalone (but not backported to master), so it would be interesting to try `ONLY_USE_BASS` again on windows.

I couldn't find any real difference, or do you mean https://github.com/vpinball/vpinball/blob/0a8dd3301945b3616917497aab257a0291654534/pinsound.h#L263 ?

**Comment by jsm174** [comment]
yes, instead of :

`pps->Play(volume, randompitch, pitch, pan, front_rear_fade, flags, restart);`

try:

```
			if (pps->IsWav2())
				pps->Play(volume, randompitch, pitch, pan, front_rear_fade, flags, (!usesame) ? true : restart);
			else
				pps->Play(volume, randompitch, pitch, pan, front_rear_fade, flags, restart); 
```

**Comment by droscoe** [comment]
Is there a preference of SDL or BASS?

**Comment by jsm174** [comment]
see above: [SDL_mixer](https://github.com/vpinball/vpinball/issues/759#issuecomment-1711788044)

**Comment by vbousquet** [comment]
BASS is not open source licensed, and we are already using SDL, so I would go for SDL.
It has to be revieewed before diving too much in it to be sure it will fit all the needed features.
I'm confident for the base features, the thing to be more carefully looked at are for SSF: 3D sound is always difficult and not always well supported (ideally we should not rely on it since it is not really designed for the SSF use case but do our own mixing and output per channel streams but this would imply quite some advanced work).

**Comment by toxieainc** [comment]
Second all of the concerns above.. But at least something open source and well maintained, please..

Back then i picked BASS because i knew it well, and there were not that many options that worked as good in practice.

**Comment by toxieainc** [comment]
@jsm174 i just backported your audio changes, thanks!!

**Comment by droscoe** [comment]
I'd be willing to look into this.  It won't be a fast process.  There's a lot I don't know about the sound system of VPX

**Comment by toxieainc** [comment]
Nice! Take your time, this should become a post-10.8 thing anyway.

Also take the freedom to restructure what you need, its super convoluted at the moment anyway (as it evolved/mutated over 2 decades ;)). And as vbousquet pointed out, the major hurdle will be 3D sound support (along with multi-soundboard).

**Comment by toxieainc** [comment]
One more thing that just came to mind: If you will pick SDL, please consider going directly to SDL3, as it apparently revamps the audio system, compared to SDL2.

**Comment by francisdb** [comment]
Just went through SDL2 mixer docs and it mentions this:

> You may only have one audio device open at a time;

https://wiki.libsdl.org/SDL2_mixer/Mix_OpenAudioDevice

SDL3 looks similar, static `audio_device`

https://github.com/libsdl-org/SDL_mixer/blob/2bf94b9a29abe0b6146cfa606b078d044683577f/src/mixer.c#L489C5-L489C18

If I understand it correctly this would no longer allow having a separate device for the backglass

**Comment by vbousquet** [comment]
I understand it does not support 7.1 positional sound but does it support 7.1 simple direct channel (I mean, doing the positional stuff on VPX end, which seems more logical since we have a very specific use case) ?

**Comment by francisdb** [comment]
Indeed, we could roll our own specific mixer. And if we don't use the mixer we can use multiple audio devices.
Not sure if that's something I could come up with though 😅

**Comment by vbousquet** [comment]
I don't know if this is that complicated. In the end, I would hope that it resolves to compute 4 attenuations (based on the player position, which is already defined in real world unit latest builds, allowing to apply traditional physics) and sending the same streams but with the attenuations to the 4 exciter channels. It could even prove to be easier to do than spoofing the native 3D positional of Windows like we do now (which expects the listener to be inside the speakers and not outside and above, and could be difficult to make portable).

**Comment by toxieainc** [comment]
Something else i just stumbled over by accident, maybe also sufficient for our needs?

https://miniaud.io/

**Comment by jsm174** [comment]
Wow! Very interesting!

**Comment by droscoe** [comment]
Whoa!

**Comment by toxieainc** [comment]
On top, this could be used for .ogg support: https://github.com/edubart/minivorbis

**Comment by francisdb** [comment]
> Something else i just stumbled over by accident, maybe also sufficient for our needs?
> 
> https://miniaud.io/

Any **open source** lib would be an improvement as long as the new situation is not stuck with ALSA only for linux. So preferably pipewire support with a fallback to pulseaudio/jack and as last resort alsa.

**Comment by toxieainc** [comment]
So you volunteer?  ;)

**Comment by francisdb** [comment]
> So you volunteer? ;)

Nope, I'm allergic to c/c++

**Comment by francisdb** [comment]
In case we are still looking for reasons to migrate to something else:
Related to SSF issues when using ogg files.

![image](https://github.com/user-attachments/assets/6ae5e8df-0471-431b-9fb3-375e00106522)

**Comment by francisdb** [comment]
Can this be closed now #2213 has been merged?

**Comment by vbousquet** [comment]
Closing as sound code has been moved to SDL / MiniAudio
